cmd/ssh.go

```diff
@@ -201,195 +202,262 @@ func (cmd *SSHCmd) startProxyTunnel(
+		if cmd.Stdio {
+			if cmd.SSHKeepAliveInterval != DisableSSHKeepAlive {
+				go startSSHKeepAlive(ctx, outerSshClient, cmd.SSHKeepAliveInterval, log)
+			}
 
-	return loftConfig, nil
-}
+			go func() {
+				if err := cmd.startRunnerServices(ctx, devPodConfig, outerSshClient, log); err != nil {
+					log.Error(err)
+				}
+			}()
 
-const loftTSNetDomain = "ts.loft"
+			go func() {
+				cmd.setupPlatformAccess(ctx, sshClient, log) // FIXME: for whatever reason this doesn't work but when running command from inside the container it works just fine
+			}()
 
-// StartSSHSession establishes an SSH session over tsnet.
-func (cmd *SSHCmd) StartSSHSession(
-	ctx context.Context,
-	ts tailscale.TSNet,
-	addr, username, password string,
-	devPodConfig *config.Config,
-	workspaceClient client2.BaseWorkspaceClient,
-	log log.Logger,
-) error {
-	workdir := filepath.Join("/workspaces", workspaceClient.Workspace())
-	if cmd.WorkDir != "" {
-		workdir = cmd.WorkDir
-	}
+			return tailscale.DirectTunnel(ctx, network, host, 8023, os.Stdin, os.Stdout)
+		}
 
-	clientConfig := &ssh.ClientConfig{
-		User: username,
-		Auth: []ssh.AuthMethod{
-			ssh.Password(password), // FIXME - use keys
-		},
-		HostKeyCallback: ssh.InsecureIgnoreHostKey(), // FIXME
-	}
+		// Connect to the inner server and handle user session
+		return machine.StartTailscaleSSHSession(
+			ctx,
+			network,
+			host,
+			cmd.User,
+			cmd.Command,
+			cmd.AgentForwarding,
+			os.Stderr,
+			sshClient,
+			envVars,
+		)
+	})
+}
 
-	// Dial the SSH server through TSNet
-	serverAddress := fmt.Sprintf("%v.%v:8022", addr, loftTSNetDomain) // FIXME
-	conn, err := ts.Dial(ctx, "tcp", serverAddress)
+// establishOuterTunnel establishes the root-level connection.
+func establishOuterTunnel(ctx context.Context, network tailscale.TSNet, host string, log log.Logger, onReady func(*ssh.Client) error) error {
+	log.Debugf("Estabilishing outer connection")
+	address := tailscale.GetURL(host, 8022)
+	conn, err := network.Dial(ctx, "tcp", address)
 	if err != nil {
-		return fmt.Errorf("failed to connect to %s: %w", serverAddress, err)
+		return fmt.Errorf("failed to connect to outer tunnel: %w", err)
 	}
 	defer conn.Close()
 
-	// Establish an SSH client connection
-	sshConn, channels, requests, err := ssh.NewClientConn(conn, serverAddress, clientConfig)
+	sshConfig := &ssh.ClientConfig{
+		User:            "root",
+		Auth:            []ssh.AuthMethod{},
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+	clientConn, chans, reqs, err := ssh.NewClientConn(conn, address, sshConfig)
 	if err != nil {
 		return fmt.Errorf("failed to establish SSH connection: %w", err)
 	}
-	client := ssh.NewClient(sshConn, channels, requests)
+	client := ssh.NewClient(clientConn, chans, reqs)
 	defer client.Close()
 
-	// Start an SSH session
-	session, err := client.NewSession()
+	log.Debugf("Outer connection ready")
+	return onReady(client)
+}
+
+// startInnerServer starts the user-level SSH server.
```

Comment:
what do you guys think about dropping the inner and outer terminology? I'm not sure how relevant it is using tailscale, we could simply have tailscaleTunnel and sshServer?

Reply:
yeah definitely, now it seems just super forced. I did this to have a better grasp of transition and do it step by step. Now we don't really have anything inner or outer, we have 2 different clients with different sets of privileges, creating sshServer is just one of the steps we execute using one of them. This can all be simplified, just wanted first to discuss how we want to proceed with those user servers in general - since now we use network ports we either want to always have a single one or we need to have some pool of ports etc.

Reply:
Awesome sounds good, yes definitely it has helped review the PR as well while we transition, but glad we are thinking along the same lines :)